src/daemon/distribution.c

```diff
@@ -0,0 +1,305 @@
+                                             double base) {
+  if ((num_buckets == 0) || (factor <= 0) || (base <= 1)) {
+    errno = EINVAL;
+    return NULL;
+  }
+
+  distribution_t *new_distribution = calloc(1, sizeof(distribution_t));
+  bucket_t *buckets = calloc(num_buckets, sizeof(bucket_t));
+
+  if ((new_distribution == NULL) || (buckets == NULL)) {
+    free(new_distribution);
+    free(buckets);
+    return NULL;
+  }
+  new_distribution->buckets = buckets;
+
+  for (size_t i = 0; i < num_buckets; i++) {
+
+    if (i == 0) {
+      new_distribution->buckets[i] = initialize_bucket(0, factor);
+    } else if (i < num_buckets - 1) {
+      new_distribution->buckets[i] = initialize_bucket(
+          new_distribution->buckets[i - 1].max_boundary, factor * pow(base, i));
+    } else {
+      new_distribution->buckets[i] = initialize_bucket(
+          new_distribution->buckets[i - 1].max_boundary, INFINITY);
+    }
+  }
+
+  new_distribution->num_buckets = num_buckets;
+  new_distribution->total_scalar_count = 0;
+  new_distribution->raw_data_sum = 0;
+  return new_distribution;
+}
+
+distribution_t *distribution_new_custom(size_t num_bounds,
+                                        double *custom_max_boundaries) {
+
+  if ((num_bounds == 0) || (custom_max_boundaries == NULL)) {
+    errno = EINVAL;
+    return NULL;
+  }
+
+  for (size_t i = 1; i < num_bounds; i++) {
+    if ((custom_max_boundaries[i] <= custom_max_boundaries[i - 1]) ||
+        (custom_max_boundaries[i] == INFINITY)) {
+      errno = EINVAL;
+      return NULL;
+    }
+  }
+
+  distribution_t *new_distribution = calloc(1, sizeof(distribution_t));
+  bucket_t *buckets =
+      calloc(num_bounds + 1, sizeof(bucket_t)); //+1 for infinity bucket
+
+  if ((new_distribution == NULL) || (buckets == NULL)) {
+    free(new_distribution);
+    free(buckets);
+    return NULL;
+  }
+  new_distribution->buckets = buckets;
+
+  for (size_t i = 0; i < num_bounds + 1; i++) {
+    if (i == 0) {
+      new_distribution->buckets[i] =
+          initialize_bucket(0, custom_max_boundaries[0]);
+    } else if (i < num_bounds) {
+      new_distribution->buckets[i] =
+          initialize_bucket(new_distribution->buckets[i - 1].max_boundary,
+                            custom_max_boundaries[i]);
+    } else {
+      new_distribution->buckets[i] = initialize_bucket(
+          new_distribution->buckets[i - 1].max_boundary, INFINITY);
+    }
+  }
+
+  new_distribution->num_buckets =
+      num_bounds + 1; // plus one for infinity bucket
+  new_distribution->total_scalar_count = 0;
+  new_distribution->raw_data_sum = 0;
+  return new_distribution;
+}
+
+static size_t binary_search(distribution_t *dist, size_t left, size_t right,
+                            double gauge) {
+  if (left > right) {
+    return -1;
+  }
+
+  size_t mid = left + (right - left) / 2;
+  if (gauge >= dist->buckets[mid].min_boundary &&
+      gauge < dist->buckets[mid].max_boundary) {
+    return mid;
+  }
```

Comment:
Thank you for your suggestion Svetlana. I added a minimal boundary attribute to the buckets for readability and practicality reasons since the buckets are non-cumulative and I will think about if it is better to remove them.